core/trino-main/src/main/java/io/trino/execution/SqlTaskExecution.java

```diff
@@ -619,30 +582,68 @@ public Driver createDriver(DriverContext driverContext, @Nullable ScheduledSplit
         {
             Driver driver = driverFactory.createDriver(driverContext);
 
-            // record driver so other threads add unpartitioned sources can see the driver
-            // NOTE: this MUST be done before reading unpartitionedSources, so we see a consistent view of the unpartitioned sources
-            drivers.add(new WeakReference<>(driver));
-
             if (partitionedSplit != null) {
                 // TableScanOperator requires partitioned split to be added before the first call to process
                 driver.updateSplitAssignment(new SplitAssignment(partitionedSplit.getPlanNodeId(), ImmutableSet.of(partitionedSplit), true));
             }
 
-            // add unpartitioned sources
-            Optional<PlanNodeId> sourceId = driver.getSourceId();
-            if (sourceId.isPresent()) {
-                SplitAssignment splitAssignment = unpartitionedSplitAssignments.get(sourceId.get());
-                if (splitAssignment != null) {
-                    driver.updateSplitAssignment(splitAssignment);
-                }
-            }
-
             status.decrementPendingCreation(pipelineContext.getPipelineId());
             closeDriverFactoryIfFullyCreated();
 
+            if (driverFactory.getSourceId().isPresent() && partitionedSplit == null) {
+                driverReferences.add(new WeakReference<>(driver));
+                scheduleSplits();
+            }
+
             return driver;
         }
 
+        public void enqueueSplits(Set<ScheduledSplit> splits, boolean noMoreSplits)
+        {
+            verify(driverFactory.getSourceId().isPresent(), "not a source driver");
+            queuedSplits.addAll(splits);
+            this.noMoreSplits.set(noMoreSplits);
```

Comment:
`verify(this.noMoreSplits.compareAndSet(false, noMoreSplits), "noMoreSplits already set");`?

Reply:
I believe the `noMoreSplits` signal may be potentially sent more then once by coordinator. Not sure if we need to be restrictive here as it doesn't impact correctness.

Reply:
Yeah - actually what I had in mind (which verify which I suggested does not do) was to check if we are not switching from this.noMoreSplits==true to this.noMoreSplits==false.

Maybe just `verify(!this.noMoreSplits || noMoreSplits, "cannot unset noMoreSplits")`